### PR TITLE
Ad hoc highlight groups

### LINKED ISSUE #17: Way to add new highlight groups?
Is there a way to add new highlighting groups, on a more-or-less ad-hoc basis? I tried using `:highlight CrystallineDirName guibg=#262626` before and after my statusline function, and while the highlight groups seem to be created correctly, they are all 'empty' (i.e. `:highlight CrystallineDirName` shows `CrystallineDirName xxx cleared`).

Can you point me in the right direction? I'm not sure *where* in the plugin code I'm supposed to look for clearing highlights…

### PR DESCRIPTION
**Pull request by runiq**
This PR allows overriding a theme's highlight groups.

### CODE REVIEW

**Comment by rbong** [comment]
So I haven't done a **super** deep dive into this yet, but this will need performance benchmarks, since highlight groups account for pretty much all of the startup time of crystalline and I'm not sure how fast `hlexists` is, and I am not sure how this will perform if `generate_sep_hi` is called (if this will even be called regularly).

**Comment by rbong** [comment]
Woops, forgot to mention - you can get the script I use for performance benchmarks [here](https://github.com/rbong/vim-crystalline/wiki/Performance-Comparison#the-test-script), and you may want to modify it.

**Comment by runiq** [comment]
So I'm getting these pretty consistently:

Benchmark (x1000) | Startup | Redraw
-|-|-
crystalline | 0:07.81 | 0:01.89
ad-hoc-highlight-groups |0:07.82 | 0:01.90

<details>
<summary>Additional information</summary>

Terminal: alacritty

Vim:
<pre>
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Mar  3 2020 00:00:00)
Included patches: 1-348
Modified by <bugzilla@redhat.com>
Compiled by <bugzilla@redhat.com>
Huge version without GUI.  Features included (+) or not (-):
+acl               -farsi             -mouse_sysmouse    -tag_old_static
+arabic            +file_in_path      +mouse_urxvt       -tag_any_white
+autocmd           +find_in_path      +mouse_xterm       -tcl
+autochdir         +float             +multi_byte        +termguicolors
-autoservername    +folding           +multi_lang        +terminal
-balloon_eval      -footer            -mzscheme          +terminfo
+balloon_eval_term +fork()            +netbeans_intg     +termresponse
-browse            +gettext           +num64             +textobjects
++builtin_terms    -hangul_input      +packages          +textprop
+byte_offset       +iconv             +path_extra        +timers
+channel           +insert_expand     +perl/dyn          +title
+cindent           +job               +persistent_undo   -toolbar
-clientserver      +jumplist          +popupwin          +user_commands
-clipboard         +keymap            +postscript        +vartabs
+cmdline_compl     +lambda            +printer           +vertsplit
+cmdline_hist      +langmap           +profile           +virtualedit
+cmdline_info      +libcall           -python            +visual
+comments          +linebreak         +python3/dyn       +visualextra
+conceal           +lispindent        +quickfix          +viminfo
+cryptv            +listcmds          +reltime           +vreplace
+cscope            +localmap          +rightleft         +wildignore
+cursorbind        +lua/dyn           +ruby/dyn          +wildmenu
+cursorshape       +menu              +scrollbind        +windows
+dialog_con        +mksession         +signs             +writebackup
+diff              +modify_fname      +smartindent       -X11
+digraphs          +mouse             -sound             -xfontset
-dnd               -mouseshape        +spell             -xim
-ebcdic            +mouse_dec         +startuptime       -xpm
+emacs_tags        +mouse_gpm         +statusline        -xsmp
+eval              -mouse_jsbterm     -sun_workshop      -xterm_clipboard
+ex_extra          +mouse_netterm     +syntax            -xterm_save
+extra_search      +mouse_sgr         +tag_binary
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/etc"
 f-b for $VIMRUNTIME: "/usr/share/vim/vim82"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H     -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -L. -Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,--enable-new-dtags -Wl,-z,relro -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld  -Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -lpython3.7m -lcrypt -lpthread -ldl  -lutil -lm  -L/usr/local/lib -Wl,--as-needed -o vim        -lm  -lselinux  -lncurses -lacl -lattr -lgpm -ldl   -Wl,--enable-new-dtags -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -fstack-protector-strong -L/usr/local/lib  -L/usr/lib64/perl5/CORE -lperl -lpthread -lresolv -ldl -lm -lcrypt -lutil -lc
</pre></details>

> since highlight groups account for pretty much all of the startup time of crystalline and I'm not sure how fast `hlexists` is

Appears to be plenty fast. If you want, I can add some additional benchmarks – I'm currently using the same vimrc for my branch as for crystalline upstream master.

> and I am not sure how this will perform if `generate_sep_hi` is called (if this will even be called regularly).

`generate_sep_hi()` (and, transitively, `synIDattrs()`) is called once per newly generated separator highlight group.

This is what I was more worried about, but at least in this benchmark my worries appear to be unfounded. Again, if you want me to add more benchmarks, I've [frankensteined the benchmark script enough](https://github.com/rbong/vim-crystalline/pull/25/commits/fcb6c21b48e0007322293f84fbec07ee6f5f4ddb) that it's not a lot of trouble anymore.

**Comment by runiq** [comment]
Rebased onto current master now that #21 has been merged. Looks pretty tiny now.

File: autoload/crystalline.vim [hunk 1]
```diff
@@ -396,6 +396,13 @@ function! crystalline#generate_theme(theme) abort
   let g:crystalline_tab_type_fake_separators = []
 
   for [l:group, l:attr] in items(a:theme)
+    if hlexists('Crystalline' . l:group)
```
Comment (rbong) [inline autoload/crystalline.vim:399]: Hmmm... same problem as in the last MR with `hlexists` here. if `'Crystalline' . l:group` has ever been defined, it won't be added here. I think that means that you can't change the theme at all if you check using `hlexists`.

**Comment by runiq** [comment]
Thinking about this after yesterday's dicussion, I just realized that this PR is unnecessary: I can override a theme's highlight groups in the `CrystallineSetTheme` autocmd.

I think I'll just close this PR and #17; the benefit would be tiny and it's already easy enough to create my own theme.

File: autoload/crystalline.vim (hunk 1 above)
Comment (runiq) [reply autoload/crystalline.vim:399]: See https://github.com/rbong/vim-crystalline/pull/22#issuecomment-603820010

**Pull request closed without merging**